common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala

```diff
@@ -0,0 +1,408 @@
+            val swapped = trySet(expectedState, WriteInProgress)
+            if (swapped) {
+                value map { _.set(new_value) }
+                transid.set(new_transid)
+            }
+            swapped
+        }
+
+        def grabInvalidationLock() = state.set(InvalidateInProgress)
+    }
+
+    /**
+     * This method posts a delete to the backing store, and either directly invalidates the cache entry
+     * or informs any outstanding transaction that it must invalidate the cache on completion.
+     */
+    protected def cacheInvalidate[R](key: Any, invalidator: => Future[R])(
+        implicit ec: ExecutionContext, transid: TransactionId, logger: Logging): Future[R] = {
+        cacheEnabled match {
+            case true =>
+                logger.info(this, s"invalidating $key")
+
+                // try inserting our desired entry...
+                val desiredEntry = new Entry(transid, InvalidateInProgress, None)
+                cache(key)(desiredEntry) flatMap { actualEntry =>
+                    // ... and see what we get back
+                    val currentState = actualEntry.state.get
+
+                    currentState match {
+                        case Cached =>
+                            // nobody owns the entry, forcefully grab ownership
+                            // note: if a new cache lookup is received while
+                            // the invalidator has not yet completed (and hence the actual entry
+                            // removed from the cache), such lookup operations will still be able
+                            // to return the value that is cached, and this is acceptable (under
+                            // the eventual consistency model) as long as such lookups do not
+                            // mutate the state of the cache to violate the invalidation that is
+                            // about to occur (this is eventually consistent and NOT sequentially
+                            // consistent since the cache lookup and the setting of the
+                            // InvalidateInProgress bit are not atomic
+                            invalidateEntryAfter(invalidator, key, actualEntry)
+
+                        case ReadInProgress | WriteInProgress =>
+                            if (actualEntry.trySet(currentState, InvalidateWhenDone)) {
+                                // then the pre-existing owner will take care of the invalidation
+                                invalidator
+                            } else {
+                                // the pre-existing reader or writer finished and so must
+                                // explicitly invalidate here
+                                invalidateEntryAfter(invalidator, key, actualEntry)
+                            }
+
+                        case InvalidateInProgress =>
+                            if (actualEntry.transid == transid) {
```

Comment:
it will be a bug if they're not propagated correctly through the system.
your observation is right.
if we don't use this transid, we'll have to create or use another id (object ref/hashcode?).